src/providers/wfs/qgswfsshareddata.cpp

```diff
@@ -466,6 +477,35 @@ bool QgsWFSSharedData::createCache()
     return false;
   }
 
+  // The id_cache should be generated once for the lifetime of QgsWFSConstants
+  // to ensure consistency of the ids returned to the user.
+  if ( mCacheIdDbname.isEmpty() )
+  {
+    mCacheIdDbname = QDir( cacheDirectory ).filePath( QStringLiteral( "wfs_id_cache_%1.sqlite" ).arg( tmpCounter ) );
+    Q_ASSERT( !QFile::exists( mCacheIdDbname ) );
+    if ( mCacheIdDb.open( mCacheIdDbname ) != SQLITE_OK )
+    {
+      QgsMessageLog::logMessage( tr( "Cannot create temporary id cache" ), tr( "WFS" ) );
+      return false;
+    }
+    QString errorMsg;
+    bool ok = mCacheIdDb.exec( QStringLiteral( "PRAGMA synchronous=OFF" ), errorMsg ) == SQLITE_OK;
+    // WAL is needed to avoid reader to block writers
+    ok &= mCacheIdDb.exec( QStringLiteral( "PRAGMA journal_mode=WAL" ), errorMsg ) == SQLITE_OK;
+    // gmlid is the gmlid or fid attribute coming from the GML GetFeature response
+    // qgisId is the feature id of the features returned to QGIS. That one should remain the same for a given gmlid even after a layer reload
+    // dbId is the feature id of the Spatialite feature in mCacheDataProvider. It might change for a given gmlid after a layer reload
+    ok &= mCacheIdDb.exec( QStringLiteral( "CREATE TABLE id_cache(gmlid TEXT, dbId INTEGER, qgisId INTEGER)" ), errorMsg ) == SQLITE_OK;
+    ok &= mCacheIdDb.exec( QStringLiteral( "CREATE INDEX idx_gmlid ON id_cache(gmlid)" ), errorMsg ) == SQLITE_OK;
+    ok &= mCacheIdDb.exec( QStringLiteral( "CREATE INDEX idx_dbId ON id_cache(dbId)" ), errorMsg ) == SQLITE_OK;
+    ok &= mCacheIdDb.exec( QStringLiteral( "CREATE INDEX idx_qgisId ON id_cache(qgisId)" ), errorMsg ) == SQLITE_OK;
+    if ( !ok )
+    {
+      QgsDebugMsg( errorMsg );
```

Comment:
I wonder if this should be more prominent (and logged to the message panel instead)

Reply:
I don't anticipate this error to happen in practice. debug is probably fine for now